### PR TITLE
MAYA-125745 - Implementation of UFE Material Interface

### PR DESCRIPTION
**Pull request by stefanminning-autodesk**
This PR implements the newly added UFE Material Interface, mainly intended for internal use in the LookdevX team.

The implemented functionality provides access to UsdSceneItems' materials, which was previously unavailable to us.

### CODE REVIEW

**Review by feldstj — DISMISSED** [review]
LGTM! Just one super minor picky point.

File: lib/mayaUsd/ufe/UsdMaterial.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    UsdMaterial(const UsdSceneItem::Ptr& item);
+    ~UsdMaterial() override;
+
+    // Delete the copy/move constructors assignment operators.
+    UsdMaterial(const UsdMaterial&) = delete;
+    UsdMaterial& operator=(const UsdMaterial&) = delete;
+    UsdMaterial(UsdMaterial&&) = delete;
+    UsdMaterial& operator=(UsdMaterial&&) = delete;
+
```
Comment (hodoulp) [inline lib/mayaUsd/ufe/UsdMaterial.h:36]: For the Doxygen documentation of the previous group of methods, the syntax is:
```
//@{
///  Delete the copy/move constructors assignment operators.
<...>
//@}
```

File: lib/mayaUsd/ufe/UsdMaterialHandler.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+//------------------------------------------------------------------------------
+// UsdMaterialHandler overrides
+//------------------------------------------------------------------------------
+
+Ufe::Material::Ptr UsdMaterialHandler::material(const Ufe::SceneItem::Ptr& item) const
+{
+    UsdSceneItem::Ptr usdItem = std::dynamic_pointer_cast<UsdSceneItem>(item);
+#if !defined(NDEBUG)
```
Comment (hodoulp) [inline lib/mayaUsd/ufe/UsdMaterialHandler.cpp:38]: The directive is useless as the `assert` is gone in release mode. Anyway the code must return if `usdItem` is null.

File: lib/mayaUsd/ufe/UsdMaterial.cpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+std::vector<PXR_NS::UsdPrim> UsdMaterial::getConnectedShaders(const PXR_NS::UsdShadeOutput& port) const
+{
+    // Dig down across NodeGraph boundaries until a surface shader is found.
+    std::vector<PXR_NS::UsdPrim> shaders;
+    for (const auto& sourceInfo : port.GetConnectedSources()) {
+        if (sourceInfo.source.GetPrim().GetTypeName() != "Shader") {
+            if (sourceInfo.sourceType == UsdShadeAttributeType::Output) {
+                const std::vector<PXR_NS::UsdPrim> connectedShaders
+                    = getConnectedShaders(sourceInfo.source.GetOutput(sourceInfo.sourceName));
+                shaders.insert(end(shaders), begin(connectedShaders), end(connectedShaders));
```
Comment (hodoulp) [inline lib/mayaUsd/ufe/UsdMaterial.cpp:52]: For `end` and `begin` the code should be `std::end` and `std::begin`. Why the compilation succeeds?

File: lib/mayaUsd/ufe/UsdMaterialHandler.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    typedef std::shared_ptr<UsdMaterialHandler> Ptr;
+
+    UsdMaterialHandler();
+    ~UsdMaterialHandler() override;
+
+    // Delete the copy/move constructors assignment operators.
+    UsdMaterialHandler(const UsdMaterialHandler&) = delete;
+    UsdMaterialHandler& operator=(const UsdMaterialHandler&) = delete;
+    UsdMaterialHandler(UsdMaterialHandler&&) = delete;
+    UsdMaterialHandler& operator=(UsdMaterialHandler&&) = delete;
```
Comment (hodoulp) [inline lib/mayaUsd/ufe/UsdMaterialHandler.h:34]: Same comment here concerning the Doxygen syntax.

**Review by hodoulp — COMMENTED** [review]
LGTM. Only few minor comments.

File: lib/mayaUsd/ufe/CMakeLists.txt [hunk 5]
```diff
@@ -96,6 +96,20 @@ if (UFE_LIGHTS_SUPPORT)
     )
 endif()
 
+message(STATUS "UFE_MATERIALS_SUPPORT is ${UFE_MATERIALS_SUPPORT}")
```
Comment (JGamache-autodesk) [inline lib/mayaUsd/ufe/CMakeLists.txt:99]: This printf might not be necessary anymore.

File: lib/mayaUsd/ufe/UsdMaterial.cpp (hunk 3 above)
Comment (JGamache-autodesk) [reply lib/mayaUsd/ufe/UsdMaterial.cpp:52]: Looks like a case of [Argument-dependent Lookup](https://en.cppreference.com/w/cpp/language/adl). The namespace of std::map::insert ca be searched for begin/end functions.

File: lib/mayaUsd/ufe/UsdMaterial.cpp [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    std::vector<PXR_NS::UsdPrim> materialPrims;
+
+    const PXR_NS::UsdPrim&                            prim = fItem->prim();
+    PXR_NS::UsdShadeMaterialBindingAPI                bindingApi(prim);
+    PXR_NS::UsdShadeMaterialBindingAPI::DirectBinding directBinding = bindingApi.GetDirectBinding();
+
+    // 1. Simple case: A material is directly attached to our object.
+    const PXR_NS::UsdShadeMaterial material = directBinding.GetMaterial();
+    if (material) {
+        materialPrims.push_back(material.GetPrim());
```
Comment (JGamache-autodesk) [inline lib/mayaUsd/ufe/UsdMaterial.cpp:82]: That is a UsdShadeMaterial. You must iterate GetSurfaceOuputs and getConnectedShaders to get the same type as will be found on geom subsets.

File: lib/mayaUsd/ufe/UsdMaterial.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    //! Create a UsdMaterial.
+    static UsdMaterial::Ptr create(const UsdSceneItem::Ptr& item);
+
+    std::vector<Ufe::SceneItem::Ptr> getMaterials() const override;
+
+private:
+    std::vector<PXR_NS::UsdPrim> getConnectedShaders(const PXR_NS::UsdShadeOutput& port) const;
```
Comment (JGamache-autodesk) [inline lib/mayaUsd/ufe/UsdMaterial.h:43]: That function does not use fItem, so it could possibly be moved to an unnamed namespace in the .cpp file.

File: lib/mayaUsd/ufe/UsdMaterialHandler.cpp (hunk 2 above)
Comment (JGamache-autodesk) [reply lib/mayaUsd/ufe/UsdMaterialHandler.cpp:38]: The equivalent in USD land is TF_AXIOM. Grep a bit to see how it is used. Requires the directive to use the PXR namespace.

File: plugin/adsk/scripts/USDMenuProc.mel [hunk 8]
```diff
@@ -76,6 +76,14 @@ global proc USDMenuProc(string $parent, string $obj)
         }
         menuItem -label "Duplicate As Maya Data" -command ("{waitCursor -state 1; mayaUsdMenu_duplicate \"" + $obj + "\"; waitCursor -state 0;}");
     }
+
+    // Allow LookdevX to define a proc with additional menu items
```
Comment (JGamache-autodesk) [inline plugin/adsk/scripts/USDMenuProc.mel:80]: Not clean. Requires secret functions shared with a third party.
Please create a publicly known function to register additional menu procs, then use the list of registered procs here. The third party will then be able to register a proc without MayaUSD having to know about it.

**Review by JGamache-autodesk — CHANGES_REQUESTED** [review]

File: lib/mayaUsd/ufe/CMakeLists.txt (hunk 5 above)
Comment (seando-adsk) [reply lib/mayaUsd/ufe/CMakeLists.txt:99]: Correct, the message is output by FindUFE.cmake

File: lib/mayaUsd/ufe/UsdMaterial.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+private:
+    UsdSceneItem::Ptr fItem;
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/UsdMaterial.h:46]: The [coding standard](https://github.com/Autodesk/maya-usd/blob/dev/doc/codingGuidelines.md) for MayaUsd is to use _item

File: lib/mayaUsd/ufe/UsdMaterial.cpp [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        } else {
+            shaders.push_back(sourceInfo.source.GetPrim());
+        }
+    }
+    return shaders;
+}
+
+std::vector<Ufe::SceneItem::Ptr> UsdMaterial::getMaterials() const
+{
+    PXR_NAMESPACE_USING_DIRECTIVE
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/UsdMaterial.cpp:63]: Why is this here? I see you are correctly using PXR_NS

File: lib/mayaUsd/ufe/UsdMaterial.h [hunk 11]
```diff
@@ -0,0 +1,50 @@
+// ===========================================================================
+// Copyright 2022 Autodesk, Inc. All rights reserved.
+//
+// Use of this software is subject to the terms of the Autodesk license
+// agreement provided at the time of installation or download, or which
+// otherwise accompanies this software in either electronic or hard copy form.
+// ===========================================================================
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/UsdMaterial.h:7]: Every file needs a license notice and copyright. Check any of the newer files for an example.

File: plugin/adsk/scripts/USDMenuProc.mel (hunk 8 above)
Comment (seando-adsk) [reply plugin/adsk/scripts/USDMenuProc.mel:80]: This was code added to follow an existing style already used in Maya to call a user defined menu proc. What Jerry is suggesting is better for what you are trying to add which is a 3rd party (not user).

**Review by seando-adsk — CHANGES_REQUESTED** [review]

File: lib/mayaUsd/ufe/UsdMaterialHandler.cpp [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    assert(usdItem);
```
Comment (ppt-adsk) [inline lib/mayaUsd/ufe/UsdMaterialHandler.cpp:39]: Better to use TF_VERIFY and return nullptr.  See
https://github.com/Autodesk/maya-usd/blob/42854b32c9c1cd706917e088b477119ca1d791ab/lib/mayaUsd/ufe/UsdShaderNodeDefHandler.cpp#L50

**Review by ppt-adsk — CHANGES_REQUESTED** [review]
Needs a test.

**Comment by ppt-adsk** [comment]
Hi @stefanminning-autodesk, can you take me off this code review?  Thanks!

**Review by seando-adsk — DISMISSED** [review]
Thanks for the test. In the future once a review has started and has comments, please don't squash. We decided this was not ideal since we lose the code review history in the pull-request (comments are no longer attached to commits since the commits don't exist anymore).

File: lib/mayaUsd/ufe/CMakeLists.txt (hunk 5 above)
Comment (JGamache-autodesk) [reply lib/mayaUsd/ufe/CMakeLists.txt:99]: Might be a good idea if you have the time.

File: plugin/adsk/scripts/USDMenuProc.mel (hunk 8 above)
Comment (JGamache-autodesk) [reply plugin/adsk/scripts/USDMenuProc.mel:80]: Actually `USDUserMenuProc` was added specifically for the needs of LookdevX (and also other third parties, as seen in #1773), so my suggestion is that LookdevX should use `USDUserMenuProc` until we get news about another plugin hijacking it.

MAYA-125745 should be about finding a way to make that extensible enough that two plugins can register their menu hooks simultaneously.

**Review by JGamache-autodesk — CHANGES_REQUESTED** [review]

**Review by JGamache-autodesk — APPROVED** [review]

**Review by seando-adsk — APPROVED** [review]
